README.md
# rcgen

[![docs](https://docs.rs/rcgen/badge.svg)](https://docs.rs/rcgen)
[![crates.io](https://img.shields.io/crates/v/rcgen.svg)](https://crates.io/crates/rcgen)
[![dependency status](https://deps.rs/repo/github/est31/rcgen/status.svg)](https://deps.rs/repo/github/est31/rcgen)

Simple Rust library to generate X.509 certificates.

```Rust
extern crate rcgen;
use rcgen::generate_simple_self_signed;
let subject_alt_names = vec!["hello.world.example".to_string(),
	"localhost".to_string()];

let cert = generate_simple_self_signed(subject_alt_names).unwrap();
// The certificate is now valid for localhost and the domain "hello.world.example"
println!("{}", cert.serialize_pem().unwrap());
println!("{}", cert.serialize_private_key_pem());
```

## Trying it out with openssl

You can do this:

```
cargo run
openssl x509 -in certs/cert.pem -text -noout
```

For debugging, pasting the PEM formatted text
to [this](https://lapo.it/asn1js/) service is very useful.

## Trying it out with quinn

You can use rcgen together with the [quinn](https://github.com/djc/quinn) crate.
The whole set of commands is:
```
cargo run
cd ../quinn
cargo run --example server -- --cert ../rcgen/certs/cert.pem --key ../rcgen/certs/key.pem ./
cargo run --example client -- --ca ../rcgen/certs/cert.der https://localhost:4433/README.md

```

## MSRV

The MSRV policy is to strive for supporting 6-month old Rust versions.

### License
[license]: #license

This crate is distributed under the terms of both the MIT license
and the Apache License (Version 2.0), at your option.

See [LICENSE](LICENSE) for details.

#### License of your contributions

Unless you explicitly state otherwise, any contribution intentionally submitted for
inclusion in the work by you, as defined in the Apache-2.0 license,
shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md

# Changes

## Release 0.11.4 - October 18, 2023

- Rename `RcGenError` to `Error` to avoid stuttering when used fully-qualified via `rcgen::`.
- Upgrade to `ring` `v0.17`.

## Release 0.11.3 - October 1, 2023

- Fix for import errors building without the optional `pem` feature.

## Release 0.11.2 - September 21, 2023

- `rcgen` has joined the umbrella of the [rustls](https://github.com/rustls) organization.
- Support for retrieving signature algorithm from `KeyPair`s. Contributed by [tindzk](https://github.com/tindzk).
- Fix for writing certificate signing requests (CSRs) with custom extensions from parameters without subject alternative names.
- Support for certificate CRL distribution points extension.
- Corrected OID for `ExtendedKeyUsagePurpose::Any`. Contributed by [jgallagher](https://github.com/jgallagher).
- Support for creating certificate revocation lists (CRLs).

## Release 0.11.1 - June 17, 2023

- Make botan a dev-dependency again. Contributed by [mbrubeck](https://github.com/mbrubeck).

## Release 0.11.0 - June 15, 2023

- Parse IP-address subject alternative names. Contributed by [iamjpotts](https://github.com/iamjpotts).
- Emit platform-apropriate line endings. Contributed by [frjonsen](https://github.com/frjonsen).
- Support larger serial numbers. Contributed by [andrenth](https://github.com/andrenth).
- Parse more certificate parameters. Contributed by [andrenth](https://github.com/andrenth).
- Output `SanType::IpAddress` when calling `CertificateParams::new` or `generate_simple_self_signed`. Contributed by [rukai](https://github.com/rukai).
- Update pem to 2.0. Contributed by [koushiro](https://github.com/koushiro).

## Release 0.10.0 - September 29, 2022

- Update x509-parser to 0.14.
- Increase minimum supported Rust version to 1.58.1.
- Update edition to 2021.
- Change `IsCa` enum to have `NoCa` and `ExplicitNoCa` and `Ca(...)`. Contributed by [doraneko94](https://github.com/doraneko94).

## Release 0.9.4 - September 28, 2022

* yanked due to breaking API changes, see 0.10.0 instead.

## Release 0.9.3 - July 16, 2022

- Add a `KeyPair::serialized_der` function. Contributed by [jean-airoldie](https://github.com/jean-airoldie).

## Release 0.9.2 - February 21, 2022

- Update x509-parser to 0.13. Contributed by [matze](https://github.com/matze).

## Release 0.9.1 - February 9, 2022

- Change edition to 2018 in order to support Rust 1.53.0.

## Release 0.9.0 - February 2, 2022

- Add RemoteKeyError for usage by remote keys.
- Support non utf8 strings. Contributed by [omjadas](https://github.com/omjadas).
- Switch from chrono to time. Contributed by [connec](https://github.com/connec).
- Update edition to 2021.

## Release 0.8.14 - October 14, 2021

- Update pem to 1.0.
- Update x509-parser to 0.12.

## Release 0.8.13 - August 22, 2021

- Bugfix release to make Certificate `Send` and `Sync` again.

## Release 0.8.12 - August 22, 2021

- Use public key as default serial number. Contributed by [jpastuszek](https://github.com/jpastuszek).
- Add support for `PKCS_RSA_SHA512` and `PKCS_RSA_SHA384` signature algorithms.
- Add support for the keyUsage extension. Contributed by [jaredwolff](https://github.com/jaredwolff).
- Ability to use remote keys. Contributed by [daxpedda](https://github.com/daxpedda).

## Release 0.8.11 - April 28, 2021

- Add getters for the criticality, content, and `oid_components` of a `CustomExtension`
- Update yasna to 0.4

## Release 0.8.10 - April 15, 2021

- Implement some additional traits for some of the types. Contributed by [zurborg](https://github.com/zurborg).
- Adoption of intra-doc-links
- Addition of the ability to zero key pairs. Contributed by [didier-wenzek](https://github.com/didier-wenzek).

## Release 0.8.9 - December 4, 2020

- Switch CI to Github Actions.
- Strip nanos from `DateTime` as well. Contributed by [@trevor-crypto](https://github.com/trevor-crypto).

## Release 0.8.7 - December 1, 2020

- Turn `botan` back into a dev-dependency. Contributed by [@nthuemmel](https://github.com/nthuemmel).
- Fix signing when CA uses different signing algorithm . Contributed by [@nthuemmel](https://github.com/nthuemmel).

## Release 0.8.6 - December 1, 2020

- Add `KeyPair::from_der`
- Add botan based test to the testsuite
- Update x509-parser to 0.9. Contributed by [@djc](https://github.com/djc).
- Ability to create certificates from CSRs. Contributed by [@djc](https://github.com/djc).

## Release 0.8.5 - June 29, 2020

- Add some more `DnType`s: `OrganizationalUnitName`, `LocalityName`, `StateOrProvinceName`
- Add `remove` function to `DistinguishedName`
- Add ability to specify `NameConstraints`

## Release 0.8.4 - June 5, 2020

- Improve spec compliance in the `notBefore`/`notAfter` fields generated by using `UTCTime` if needed

## Release 0.8.3 - May 24, 2020

- Fix regression of `0.8.1` that generated standards non compliant CSRs
  and broke Go toolchain parsers. Contributed by [@thomastaylor312](https://github.com/thomastaylor312).

## Release 0.8.2 - May 18, 2020

- Disable `chrono` default features to get rid of time crate
- Improve `openssl` tests to do a full handshake with the generated cert

## Release 0.8.1 - April 2, 2020

- Fix non-standard-compliant SubjectKeyIdentifier X.509v3 extension format
- BasicConstraints X.509v3 extension is now marked as critical
- Use RFC 7093 to calculate calculate subject key identifiers
- Add option to insert AuthorityKeyIdentifier X.509v3 extension into non-self-signed certificates
- Update to x509-parser 0.7

## Release 0.8.0 - March 12, 2020

- Update to pem 0.7
- Correct number of nanoseconds per second. Contributed by [@samlich](https://github.com/samlich).
- Adoption of the `non_exhaustive` feature in the API

## Release 0.7.0 - September 14, 2019

- Bugfix release for ip address subject alternative names.
  Turns out they aren't CIDR subnets after all :)

## Release 0.6.0 - September 12, 2019

- Support for email and cidr subnet (ip address) subject alternative names

src/oid.rs
/// pkcs-9-at-extensionRequest in RFC 2985
pub const OID_PKCS_9_AT_EXTENSION_REQUEST: &[u64] = &[1, 2, 840, 113549, 1, 9, 14];

/// id-at-countryName in RFC 5280
pub const OID_COUNTRY_NAME: &[u64] = &[2, 5, 4, 6];
/// id-at-localityName in RFC 5280
pub const OID_LOCALITY_NAME: &[u64] = &[2, 5, 4, 7];
/// id-at-stateOrProvinceName in RFC 5280
pub const OID_STATE_OR_PROVINCE_NAME: &[u64] = &[2, 5, 4, 8];
/// id-at-organizationName in RFC 5280
pub const OID_ORG_NAME: &[u64] = &[2, 5, 4, 10];
/// id-at-organizationalUnitName in RFC 5280
pub const OID_ORG_UNIT_NAME: &[u64] = &[2, 5, 4, 11];
/// id-at-commonName in RFC 5280
pub const OID_COMMON_NAME: &[u64] = &[2, 5, 4, 3];

// https://tools.ietf.org/html/rfc5480#section-2.1.1
pub const OID_EC_PUBLIC_KEY: &[u64] = &[1, 2, 840, 10045, 2, 1];
pub const OID_EC_SECP_256_R1: &[u64] = &[1, 2, 840, 10045, 3, 1, 7];
pub const OID_EC_SECP_384_R1: &[u64] = &[1, 3, 132, 0, 34];

// rsaEncryption in RFC 4055
pub const OID_RSA_ENCRYPTION: &[u64] = &[1, 2, 840, 113549, 1, 1, 1];

// id-RSASSA-PSS in RFC 4055
pub const OID_RSASSA_PSS: &[u64] = &[1, 2, 840, 113549, 1, 1, 10];

// https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.3
pub const OID_KEY_USAGE: &[u64] = &[2, 5, 29, 15];

// https://tools.ietf.org/html/rfc5280#appendix-A.2
// https://tools.ietf.org/html/rfc5280#section-4.2.1.6
pub const OID_SUBJECT_ALT_NAME: &[u64] = &[2, 5, 29, 17];

// https://tools.ietf.org/html/rfc5280#section-4.2.1.9
pub const OID_BASIC_CONSTRAINTS: &[u64] = &[2, 5, 29, 19];

// https://tools.ietf.org/html/rfc5280#section-4.2.1.2
pub const OID_SUBJECT_KEY_IDENTIFIER: &[u64] = &[2, 5, 29, 14];

// https://tools.ietf.org/html/rfc5280#section-4.2.1.1
pub const OID_AUTHORITY_KEY_IDENTIFIER: &[u64] = &[2, 5, 29, 35];

// id-ce-extKeyUsage in
// https://tools.ietf.org/html/rfc5280#section-4.2.1.12
pub const OID_EXT_KEY_USAGE: &[u64] = &[2, 5, 29, 37];

// id-ce-nameConstraints in
// https://tools.ietf.org/html/rfc5280#section-4.2.1.10
pub const OID_NAME_CONSTRAINTS: &[u64] = &[2, 5, 29, 30];

// id-ce-cRLDistributionPoints in
// https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.13
pub const OID_CRL_DISTRIBUTION_POINTS: &[u64] = &[2, 5, 29, 31];

// id-pe-acmeIdentifier in
// https://www.iana.org/assignments/smi-numbers/smi-numbers.xhtml#smi-numbers-1.3.6.1.5.5.7.1
pub const OID_PE_ACME: &[u64] = &[1, 3, 6, 1, 5, 5, 7, 1, 31];

// id-ce-cRLNumber in
// https://www.rfc-editor.org/rfc/rfc5280#section-5.2.3
pub const OID_CRL_NUMBER: &[u64] = &[2, 5, 29, 20];

// id-ce-cRLReasons
// https://www.rfc-editor.org/rfc/rfc5280#section-5.3.1
pub const OID_CRL_REASONS: &[u64] = &[2, 5, 29, 21];

// id-ce-invalidityDate
// https://www.rfc-editor.org/rfc/rfc5280#section-5.3.2
pub const OID_CRL_INVALIDITY_DATE: &[u64] = &[2, 5, 29, 24];

// id-ce-issuingDistributionPoint
// https://datatracker.ietf.org/doc/html/rfc5280#section-5.2.5
pub const OID_CRL_ISSUING_DISTRIBUTION_POINT: &[u64] = &[2, 5, 29, 28];
